torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -1266,30 +1266,25 @@ Tensor infinitely_differentiable_logit_backward(
   }
 }
 
-Tensor kl_div_double_backward_grad_output(const Tensor & grad, const Tensor & input, const Tensor & target, int64_t reduction, bool log_target) {
-  auto result = kl_div_backward(grad, input, target, at::Reduction::None, log_target);
+Tensor kl_div_backward(const Tensor& grad_output, const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
+  auto grad_input = (
+    log_target ? -at::exp(target)
+               : -target
+  ) * grad_output;
   if (reduction == at::Reduction::Mean) {
-    return result.mean();
-  } else if (reduction == at::Reduction::Sum) {
-    return result.sum();
+    grad_input /= input.numel();
   }
-  return result;
+  return grad_input;
 }
 
-// Compute derivatives for targets.
-Tensor kl_div_target_backward(Tensor grad_output, Tensor self, Tensor target, int64_t reduction, bool log_target) {
-  Tensor grad_target;
-  if (!log_target) {
-    grad_target = grad_output.mul(target.log().add_(1).sub_(self)).masked_fill_(target == 0, 0.);
-  }
-  else {
-    grad_target = grad_output.mul(target.add(1).sub_(self).mul_(target.exp()));
-  }
-
+Tensor kl_div_target_backward(const Tensor& grad_output, const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
```

Comment:
Would you elaborate on the changes here, too?

Reply:
Same thing as in the other case, it concerns negative numbers, for which this function is not defined. It is also faster to do it this way, and it's also consistent with the docs.